src/common/pmix_log.c

```diff
@@ -69,13 +69,9 @@ PMIX_EXPORT pmix_status_t PMIx_Log(const pmix_info_t data[], size_t ndata,
     pmix_cb_t cb;
     pmix_status_t rc;
 
-    PMIX_ACQUIRE_THREAD(&pmix_global_lock);
```

Comment:
Hmm, that complicates things.

If prted or prte (or a call PMIx_foo()) calls show_help while already holding the global lock, it will deadlock. Perhaps we need an internal-only log function for show-help to use that doesn't lock?

Reply:
Hm, maybe we can add these checks to the thread shift call? that should be ok?

Reply:
That should never happen. We only hold the lock to check that init counter and then immediately release it. Even if two app threads call PMIx functions, we should never deadlock due to that check.

Reply:
Let me offer an alternative. The intent of the lock/check was to ensure that the client/server/tool "init" functions had been called and completed prior to allowing someone to call a PMIx API. Otherwise, the caller could begin accessing parts of PMIx before they had been initialized. At the time we did it, there was no problem calling any internal PMIx function from inside the lock because they don't check that lock - it was only done at the beginning of an API. Only exception was in the "init" functions themselves as we had to ensure they completed prior to letting someone into an API.

An aside: this is precisely why we don't lock at the object level. If I am using an object and call a function that calls a function that touches that object, then we deadlock. Very hard to ensure it doesn't happen.

The problem here, therefore, is that we are now calling functions during those "init" routines while holding the lock that could call "show_help", and (as you noted) "show_help" now checks the lock because it calls a PMIx API. Since this is specific to the "init" routines, we have a couple of choices:

* release the lock prior to calling an internal function. We might want to reorder things so that such calls occur late in the "init" routine, after we have setup all the global structures.
* remove the lock and simply put it on the user by dictating that "you cannot call a PMIx function until _after_ the respective init function completes".

I'd personally opt for the first as it is the least risky. It still opens a window for potential trouble, but if we can ensure all globals have been at least initialized, anyone calling an API should at least not segfault.

Make sense?

Reply:
I think option 1 would require a large refactor, and will be error/mistake prone. Unless you think it will be a smaller change than what I am thinking - modifying the init routines just to accommodate show_help() for error cases seems too risky for me for release branches at this point in the game for v4.2. It could be something considered for master, but I'd be hesitant to bring that to the v4.2 branch (unless, again, it is a smaller refactor than I am thinking).

Option 2 seems __ok__ for now - since as you said, users should not be calling Log before Init anyway. If I had to choose between the two for release branches, I'd probably choose that one.

A third option would be to use a [pthread_readwrite_lock](https://pubs.opengroup.org/onlinepubs/009604599/functions/pthread_rwlock_rdlock.html)(). This will allow readers of these global values (which what PMIx_Log/_nb are doing - just checking the init counters and such) to read these values, even if they hold the write lock. IE PMIx_init() will acquire the write lock in its init routine, and won't deadlock if it calls show_help() because that will only require a read_lock(). This seems like it solves both problems, init functions can safely call lock while still checking these values, and users will still get the error message/check of calling these routines before initializing PMIx. 

I'm not sure if that is safer than modifying/shuffling the init code, but I do think it will be less prone to code drift in the future in avoiding deadlock cases with this path.

Reply:
let me poke at this a bit over the weekend - we've encountered this a bit before, and i think i recall how we got around it.